memory: new memory layout

abonislawski opened the pull request:
It introduces two significant optimizations:

> Separate the memory (heaps, stack) by cores so if we will turn off a specific core we can also turn off all memory of this core.
> Remove fixed defines and gaps, previously stack was at fixed position - end of memory and buffer heap was everything between stack and fw. Now everything will be packed at sof_fw memory segment and takes only this memory which is really needed, everything else we can turn off.


BSS section now includes heaps and stack
Removed many heap fixed position defines
Removed gap before fw text section in CNL
memmap init is using real linker positions for heaps
Temporary stack for boot_ldr placed in heap_hp_buffer (safe)
use linker stack position in boot entry
use linker heap position in sanity check
sof fw max_size limited to memory configured by memory banks count
    
Using new macro to stack init per core
Stack is no longer a one big block for all cores (core_count*stack_size)
System_runtime_heap and system_heap are no longer a one big block for all cores
    
Memory layout for every slave core aligned to hw memory bank:

    --align to memory bank--
    system_heap
    system_runtime_heap
    stack
    --align to memory bank--
    next slave core..

So if we will turn off a specific core then we are also able to turn off memory "connected" to this core    

Similar layout applies to master core

Signed-off-by: Adrian Bonislawski <adrian.bonislawski@linux.intel.com>

dbaluta commented:
@abonislawski none of the commits explain why do we need a new memory layout and what's wrong with the existing one. Please fix this.

lgirdwood commented:
@abonislawski also this breaks CI.

tlauda commented on src/arch/xtensa/smp/xtos/xtos-internal.h:
```diff
@@ -333,6 +334,29 @@ XTOS_PENDING_OFS:	.space	4	/* _xtos_pending variable */
+	// Retrieves address of end of stack buffer for certain core to register ax.
+	.macro	xtos_stack_addr_percore ax, ay, stack_master, stack_slave, stack_size
+	get_prid	\ax
+	bnez		\ax, core_s
```
Core 0 doesn't need to be master core - that's why we have `PLATFORM_MASTER_CORE_ID`.

abonislawski replied:
It is a real scenario or just a theoretical option? I can branch in this macro if core is not a PLATFORM_MASTER_CORE_ID but it will work correctly only if it is 0, just like in memmap in memory.c - it is designed for master at 0

tlauda replied:
Real scenario as there is already HW with master core != 0 (although currently not supported by SOF). I know it probably won't work due to some historical assumptions, but better not to limit the new code.

mmaka1 replied:
Yes, if we should not assume master == 0 in the generic xtensa arch code since the ID of the core must be unique and that is it, the actual value depends on how the cores are integrated.

We may assume it temporarily (better not, however safe atm) in the cavs platform code since we know how the cores are integrated in there.

jajanusz commented on test/cmocka/include/mock_memory.h:
```diff
@@ -0,0 +1,6 @@
+/* Memory mock for memmap */
```
Missing license

jajanusz commented:
Doesn't it break `git bisect`? I didn't look into details, but by looking at commit titles commits like `new memory layout` should be squashed with minimal commit that changes what layout is needed. Otherwise it will not build for other platforms, for few commits, making git bisect not so useful.

abonislawski commented:
@jajanusz all memory layout commits squashed. 

One more thing, please check intel licence note in xtos-internal.h, it should be added like this? Should I add author note? And it should be also added to boot_entry.S and crt1-boards.S If I have changed only 1 line in these files?

abonislawski commented:
Because of recent changes on master this PR will work again when #1522 will be merged

tlauda commented on src/platform/apollolake/include/platform/memory.h:
```diff
@@ -152,17 +152,12 @@
 
-#define HEAP_BUFFER_BASE	(HEAP_RUNTIME_BASE + HEAP_RUNTIME_SIZE)
-#define HEAP_BUFFER_SIZE	(SOF_STACK_END - HEAP_BUFFER_BASE)
+#define HEAP_BUFFER_SIZE	0x2000
```
Buffer heap has only 8kB?

abonislawski replied:
Yes, looks like a safe and tested value, @dabekjakub can share more details

lgirdwood replied:
@abonislawski this looks too small for large buffers or pipelines with multiple buffers. @dabekjakub can you explain the details.

abonislawski replied:
@lgirdwood I have asked @wwittbrx for test it on a real platform and it works very well

tlauda replied:
@abonislawski @wwittbrx It should fail for more than 2 pipelines of 48 kHz, 8 channels, 32 bit.

abonislawski replied:
issue resolved, just a note for @lgirdwood : @tlauda noticed that there is also used a second buffer HEAP_HP_BUFFER, so no problem with 0x2000 for HEAP_BUFFER_SIZE, tested also with 8 channels@32b

abonislawski commented:
> No idea why you are making these changes are there are no commit messages and the PR description fails to explain why each change is needed.

"This will optimize memory usage, remove fixed position defines and allow to
power down all memory of a specified core" + changelog

I will add more info to commit messages if thats not enough

It introduces two significant optimizations:
1. Separate the memory (heaps, stack) by cores so if we will turn off a specific core we can also turn off all memory of this core.
2. Remove fixed defines and gaps, previously stack was at fixed position - end of memory and buffer heap was everything between stack and fw. Now everything will be packed at sof_fw memory segment and takes only this memory which is really needed, everything else we can turn off.

lgirdwood commented on src/platform/intel/cavs/memory.c:
```diff
@@ -9,6 +9,9 @@
 #include <cavs/memory.h>
 
+extern unsigned int _system_heap, _system_runtime_heap, _module_heap;
+extern unsigned int _buffer_heap, _sof_core_s_start;
```
If these are pointers we should use _t pointer type.

abonislawski commented:
@lgirdwood sure, good point. I have updated ASCII descriptions of the memory map.

Also added one commit for imx8 linker so all of them will get fix

abonislawski commented:
Force pushed because of QB issue, no changes today. Please finish your reviews @tlauda @jajanusz @mmaka1 @dbaluta

lyakh commented:
@abonislawski @lgirdwood this broke Sue Creek.

jajanusz commented:
@lyakh Sorry we have no suecreek here and we just try to not cause CI regression.
@lgirdwood Can we add something more than just building for suecreek in CI?

abonislawski commented:
@lyakh I cannot test it without platforms but PR #1567 will probably fix it, please let me know
